include/pybind11/embed.h

```diff
@@ -146,9 +115,54 @@ inline void initialize_interpreter(bool init_signal_handlers = true,
         pybind11_fail("The interpreter is already running");
     }
 
+#if PY_VERSION_HEX < 0x030B0000
+
     Py_InitializeEx(init_signal_handlers ? 1 : 0);
 
-    detail::set_interpreter_argv(argc, argv, add_program_dir_to_path);
+    // Before it was special-cased in python 3.8, passing an empty or null argv
+    // caused a segfault, so we have to reimplement the special case ourselves.
+    bool special_case = (argv == nullptr || argc <= 0);
+
+    const char *const empty_argv[]{"\0"};
+    const char *const *safe_argv = special_case ? empty_argv : argv;
+    if (special_case) {
+        argc = 1;
+    }
+
+    auto argv_size = static_cast<size_t>(argc);
+    // SetArgv* on python 3 takes wchar_t, so we have to convert.
+    std::unique_ptr<wchar_t *[]> widened_argv(new wchar_t *[argv_size]);
+    std::vector<std::unique_ptr<wchar_t[], detail::wide_char_arg_deleter>> widened_argv_entries;
+    widened_argv_entries.reserve(argv_size);
+    for (size_t ii = 0; ii < argv_size; ++ii) {
+        widened_argv_entries.emplace_back(detail::widen_chars(safe_argv[ii]));
+        if (!widened_argv_entries.back()) {
+            // A null here indicates a character-encoding failure or the python
+            // interpreter out of memory. Give up.
+            return;
+        }
+        widened_argv[ii] = widened_argv_entries.back().get();
+    }
+
+    auto *pysys_argv = widened_argv.get();
+
+    PySys_SetArgvEx(argc, pysys_argv, static_cast<int>(add_program_dir_to_path));
+#else
+    PyConfig config;
+    PyConfig_InitPythonConfig(&config);
+    config.safe_path = add_program_dir_to_path ? 0 : 1;
```

Comment:
@vstinner Do you know why this might not be working in 3.11? I've swapped the third argument of `PySys_SetArgvEx` with this setting (reversed, since it's true to remove the local directory from the search path), but 3.11 is unable to find the local modules in the tests (and then segfaults for good measure). I think we are working around https://github.com/python/cpython/issues/92678 correctly but the tests break on this.

Reply:
Okay, I think what's happening is the newer API is not adding this to the path (regardless of `safe_path`, which only affects `Py_RunMain` - the new API pre-dates safe_path). So I have to manually add the [old algorithm](https://docs.python.org/3/c-api/init.html#c.PySys_SetArgvEx) of pulling out the script directory if an existing path is the first argument, and adding an empty string otherwise.

Reply:
`_PyPathConfig_ComputeSysPath0` is private. :( It seems looking at PEP 587 that all the `sys.path` handling is now tied to the high level API (`Py_Main`, `Py_BytesMain`, and `Py_RunMain`), and line-by-line runners like us are stuck reimplementing the logic ourselves now.